Add driver for Bosch BMP180 temp/pressure sensor

fd0 opened the pull request:
This PR adds a driver for the Bosch BMP180 temperature/pressure sensor. At the moment the driver works and reads sensible data.

Closes #130

quinte17 commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+// Copyright 2016 The Periph Authors. All rights reserved.
```
is here 2017 preferable?

fd0 replied:
Sure.

quinte17 commented on cmd/bmp180/main.go:
```diff
@@ -0,0 +1,119 @@
+// Copyright 2016 The Periph Authors. All rights reserved.
```
is here 2017 preferable?

fd0 replied:
I think so.

quinte17 commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+	d.c.MB = int16(cal[16])<<8 | int16(cal[17])
+	d.c.MC = int16(cal[18])<<8 | int16(cal[19])
+	d.c.MD = int16(cal[20])<<8 | int16(cal[21])
+
```
please consider something like
`convert(cal[:], &d.c)`
where
```
import "encoding/binary"
import "bytes"

func convert(b []byte, data interface{}) error {
	buf := bytes.NewReader(b)
	return binary.Read(buf, binary.LittleEndian, data)
}
```

fd0 replied:
Hm, I disagree here:
 * Is that really necessary? I've use the same style as in the `bme280` driver, which uses exactly this here: https://github.com/google/periph/blob/master/devices/bme280/bme280.go#L257-L269
 *  It feels wrong to me. What's happening here is that the bytes are interpreted as big endian (not little endian) according to the datasheet. Having a separate function call, that allocates a bytes.Reader struct, only to read two bytes out of it feels like A LOT overhead.
 * IMHO it makes the code much harder to read.

tve replied:
To me the `convert` idea is pretty slick. It could actually be a method on `calibration` that takes the byte array as arg and returns a `calibration` struct, mhh, could actually be newCalibration. You're creating a binary.BigEndian reader for the entire struct, not for each 2-byte element. Or am I not getting it right?

maruel replied:
bme280 uses custom because of weirdo bit packing at nibble boundary for the "h" ones.

https://periph.io/x/periph/conn/mmr#Dev8.ReadStruct is what you want.

quinte17 replied:
the convert idea gets all your single lines into a oneliner..
ok maybe i missed the endianess.
example usage can be seen here:
https://github.com/quinte17/bme280/blob/master/bme280.go#L357
but i agree for bme280 its not perfect either..

fd0 replied:
@quinte17 Ah! I re-read your comment and I missed that you pass a pointer to the calibration struct to `convert()`, now it makes sense. Thanks!

maruel commented on cmd/bmp180/main.go:
```diff
@@ -0,0 +1,119 @@
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// bmp180 reads the current temperature and pressure from a BMP180.
```
/me start thinking about more coherency for package docstring
https://godoc.org/periph.io/x/periph/devices#pkg-subdirectories

At the very least, it shall start with: `Package bmp180 ...`

Include a Reference section with a link to the datasheet.
https://ae-bst.resource.bosch.com/media/_tech/media/datasheets/BST-BMP180-DS000-121.pdf

fd0 replied:
Uhm, that's not package `bmp180`, but package `main`. The package `bmp180` has a proper doc string and a link to the datasheet:
```
// Package bmp180 controls a Bosch BMP180 device over I²C.
//
// Datasheet
//
// https://cdn-shop.adafruit.com/datasheets/BST-BMP180-DS000-09.pdf
```

The datasheet you linked is really bad btw, the font in the algorithm on page 15 is unreadable. The one at adafruit that I linked has at least a readable font in the screenshot...

maruel replied:
Oops sorry I was really confused, ignore my comment about package name.

It kind of suck that the official one is of less quality. I'm not a fan of linking to adafruit's copies.

fd0 replied:
I agree. Maybe we can put both links in? I also found at least two errors in the official datasheet...

maruel replied:
Sure!

BTW sorry for the feeling of nitpicking fest; I'm trying to keep a very specific style on the library and for the device drivers I really want to behave the best w.r.t. the device (e.g. like sleeping the exact time as specified by the datasheet). I hope it's not a bummer for you.

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+	return nil
+}
+
+// Opts is optional options to pass to the constructor.
```
Since there's only one parameter, pass the parameter to New() directly. No need for a struct.

fd0 replied:
Ok.

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+
+const softResetValue = 0xB6
+
+func (d *Dev) makeDev(opts *Opts) error {
```
Since there's no SPI support (the datasheet says it's on custom orders only, so not worth implementing), then there's no need of a makeDev() function.

fd0 replied:
Ok.

maruel commented on cmd/bmp180/main.go:
```diff
@@ -0,0 +1,119 @@
+		if !loop {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
```
Use the conversion time from the datasheet at page 21. 100ms is waaay too long.

fd0 replied:
Ok

tve replied:
Since this is in a loop printing to stdout maybe it should be something slower that one can follow? Such as 2x per second?

maruel replied:
Oh right, for printing to the user, you can make this configurable through https://golang.org/pkg/flag/#Duration with `time.Second` the default. Also you want once per second, not sleep one second, which means using https://golang.org/pkg/time/#After

(I haven't done thing in bme280/main.go, it'd be a good thing to fix there too)

fd0 replied:
Fine with me.

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+)
+
+// New returns an object that communicates over I²C to BMP180 environmental
+// sensor.
```
Add a note that the device can work at up to 3.4MHz. This is valuable information.

fd0 replied:
OK.

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+}
+
+const (
+	bmp180Address = 0x77 // the address of a BMP180 is fixed at 0x77
```
I'm not a huge fan of constants that are used exactly once.

fd0 replied:
I disagree: IMHO assigning a name helps readability a lot. This gives a magic constant that is suddenly used in the code a name, which can e.g. be searched in the datasheet. This is a common practice in development for embedded systems (at least in my experience). Thoughts?

maruel replied:
You are in package bmp180, so prefixing a const with "bmp180" is redundant.

It was more of a nit than anything else, I don't mind much.

fd0 replied:
Ok, I can live with that :)

What I meant was mainly tailored towards magic constants that have been sourced from somewhere else (e.g. the datasheet).

maruel commented on devices/bmp180/bmp180_test.go:
```diff
@@ -0,0 +1,29 @@
+package bmp180
```
add standard header.

fd0 replied:
"header" as in "package header" or "copyright information"?

maruel replied:
Copyright header.

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+		return err
+	}
+
+	UT := uint16(tempBuf[0])<<8 | uint16(tempBuf[1])
```
why uppercase?

fd0 replied:
I used the syntax from the manual.

maruel replied:
still, upper case has special meaning in Go so use lower case.

fd0 replied:
Hm, I'm not aware of any special meaning of the case of variables within a function, am I missing something?

I'll use lower case.

maruel replied:
You are right there is no meaning inside a function, but the code style is to always use lower case names for locally scoped symbols.

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+		return err
+	}
+
+	time.Sleep(100 * time.Millisecond)
```
too long

fd0 replied:
Agreed.

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+	return nil
+}
+
+func (d *Dev) readReg(reg uint8, b []byte) error {
```
You may want to use periph.io/x/periph/conn/mmr which exists exactly for this.

The bme280 driver couldn't use it because of the SPI weirdo communication protocol.

fd0 replied:
Ah, thanks! I wasn't aware of the mmr package.

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+
+// calibration data read from the internal EEPROM (datasheet page 13)
+type calibration struct {
+	AC1, AC2, AC3 int16
```
why export the members? Please use lowercase even if they are uppercase in the datasheet.

fd0 replied:
Hm, I'm pretty sure nothing is exported here since the struct itself is not exported...

fd0 replied:
That doesn't work: Using reflection on not exported struct members fails, so we cannot use `binary.Read()` to parse the calibration data. I'm leaving this upper case.

Besides: The struct itself is not exported, so it's not as if it will show up somewhere, not even in godoc...

maruel replied:
Oh sure, if (and only if) you switch to using reflection, then keep them exported.

The idea here is to always aim for least privilege. Uppercase/lowercase is a way to document to the _reader of the code_ if there are interactions that falls outside of the assignments as shown in the package.

maruel commented:
gohci

maruel commented:
(I need to fix the gohci check)

maruel commented:
gohci: test

(sorry I'm using your PR as a test bed to enable gohci jobs on PRs)

tve commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+	O8x Oversampling = 3
+)
+
+// Dev is an handle to a bmp180.
```
s/an/a/

tve commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+)
+
+// Sense returns measurements as °C and kPa.
+func (d *Dev) Sense(env *devices.Environment) error {
```
@maruel why doesn't sense return devices.Environment?

maruel replied:
in case the struct grows a lot to save on allocation, I agree it could be changed, I'm not confident with this interface.

tve replied:
given everything else, the allocation of the struct seems irrelevant to me...

fd0 replied:
I agree, it's irrelevant. But I also think it is outside the scope of this PR to change the API :)

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+// bit offsets in regCtrlMeas
+const (
+	ctrlMeasurementControl = 0
+	ctrlStartConversion    = 5
```
These are bitshift, so their name should be ctrlMeasurementShift, ctrlStartConversionShift and ctrlOversampleShift. Otherwise the name is confusing.

fd0 replied:
Agreed.

maruel commented on devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,258 @@
+}
+
+const (
+	ctrlCommandTemperature = 0x0E
```
Frankly it'd make more sense if the constants were fixed here, e.g.
```
ctrlCommandTemperature = 0x0E << ctrlMeasurementControl
```

but since these commands cannot run without the start conversion bit, I don't see value in having these constant separated.

fd0 replied:
I've renamed the constants here to `cmdStartTempConv`, I think that makes more sense. Thanks for the suggestion.

fd0 commented:
I've addressed all comments and moved the device to `experimental/devices` (and I've verified that the driver still works of course). Looking forward to further comments!

maruel commented on cmd/bmp180/main.go:
```diff
@@ -0,0 +1,127 @@
+
+func mainImpl() error {
+	i2cID := flag.String("i2c", "", "I²C bus to use")
+	hz := flag.Int("hz", 0, "I²C bus speed")
```
I'd say to not bother, I had added these because I was experimenting but it's really irrelevant for a sensing device, I'll post a change to remove it from the bme280 exec.

Also setspeed is not working fine on the rpi so it's not even usable at the time. :(

fd0 replied:
OK

maruel commented on cmd/bmp180/main.go:
```diff
@@ -0,0 +1,127 @@
+	sample4x := flag.Bool("s4", false, "sample at 4x")
+	sample8x := flag.Bool("s8", false, "sample at 8x")
+	loop := flag.Bool("l", false, "read data continously")
+	interval := flag.Duration("i", time.Second, "read data continously with this interval")
```
what about just having -i, make it default to 0?
then no need for a second flag. will update bme280 later.

fd0 replied:
Good idea.

maruel commented on cmd/bmp180/main.go:
```diff
@@ -0,0 +1,127 @@
+
+	var dev *bmp180.Dev
+
+	fmt.Printf("open bmp180\n")
```
don't forget to remove

maruel commented on cmd/bmp180/main.go:
```diff
@@ -0,0 +1,127 @@
+	var dev *bmp180.Dev
+
+	fmt.Printf("open bmp180\n")
+	if dev, err = bmp180.New(bus, os); err != nil {
```
extract the assignment from the if, so no need to define dev upfront

fd0 replied:
OK

maruel commented on experimental/devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,259 @@
+	time.Sleep(10 * time.Millisecond)
+
+	// read calibration data from internal EEPROM, 11 registers with two bytes each
+	var cal [22]byte
```
cal := calibration {}

fd0 replied:
But then we cannot do the check a few lines later easily, as the datasheet suggests...

maruel replied:
The main reason I prefer this is because of the current code use reflection twice. It is mostly a bug in the API design of `mmr` but still. The fact that big endian decoding is done part manually, part automatically, is kind of weird.

```
func isValid(i int16) bool {
  return i != 0 && i != ^int16(0)
}
func isValidU(i uint16) bool {
  return i != 0 && i != 0xFFFF
}

func (c*calibration) isValid() bool {
  return isValid(c.AC1) && isValid(c.AC2) && isValid(c.AC3) && isValidU(c.AC4) && isValidU(c.AC5) && isValidU(c.AC6) && isValid(c.B1) && isValid(c.B2) && isValid(c.MB) && isValid(c.MC) && isValid(c.MD)
}
```

quinte17 replied:
should the newCalibration func validate before returning a correct instance?

fd0 replied:
Nice solution, thanks. And removing the double endian decoding is much better code.

maruel commented on experimental/devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,259 @@
+	return nil
+}
+
+const (
```
group all consts groups without functions in between.

but: I generally prefer public symbols to be first, the. private .
then each group has const, var, the code. it's not an hard requirement

fd0 commented:
Ok, more improvements, thanks!

maruel commented on cmd/bmp180/main.go:
```diff
@@ -0,0 +1,110 @@
+		if err := e.Sense(&env); err != nil {
+			return err
+		}
+		fmt.Printf("%8s %10s %9s\n", env.Temperature, env.Pressure, env.Humidity)
```
There's no humidity sensing so it should not be printed.

fd0 replied:
OK

maruel commented on experimental/devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,265 @@
+	ctrlMeasurementControlShift = 0
+	ctrlStartConversionShift    = 5
+	ctrlOversamplingShift       = 6
+)
```
optional nit: use one const () section, with one empty line being the separation between logical groups.

maruel replied:
FTR, the reason is that I care about vertical density; the more dense the code can be vertically, the more a screenful page shows more code, leading to better visualization. Obviously this is a trade off with readability, as some people need more vertical spacing than others but I feel it is generally abused so I prefer not too many empty lines.

fd0 replied:
Works for me.

maruel commented on experimental/devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,265 @@
+	return isValid(c.AC1) && isValid(c.AC2) && isValid(c.AC3) && isValidU(c.AC4) && isValidU(c.AC5) && isValidU(c.AC6) && isValid(c.B1) && isValid(c.B2) && isValid(c.MB) && isValid(c.MC) && isValid(c.MD)
+}
+
+// newCalibration parses calibration data from the buf and checks that it is valid.
```
remove now dead code

maruel commented on experimental/devices/bmp180/bmp180_test.go:
```diff
@@ -0,0 +1,33 @@
+		MD:  2868,
+	}
+
+	temp := c.compensateTemp(27898)
```
merge lines 24-25 into one
merge lines 39-40 into one

maruel commented on experimental/devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,265 @@
+	regCalibrationStart = 0xAA // first calibration register address
+	regSoftReset        = 0xE0 // soft reset register
+	regCtrlMeas         = 0xF4 // control measurement
+	regOutMSB           = 0xF6 // MSB of measurement
```
why not just `regOut = 0xF6 // 3 bytes register with measurement data`
the rest is not used directly anyway.

fd0 replied:
OK.

maruel commented on experimental/devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,265 @@
+		return nil, fmt.Errorf("bmp180: unexpected chip id 0x%x; is this a BMP180?", id)
+	}
+
+	// issue soft reset to initialize device
```
I don't see why this is needed? I'm fine with exposing `Reset()` as a method but I don't see the need to call it every time the device is instantiated.

fd0 replied:
You're right: It is not necessary (in most cases). I'd like to make sure that the device is in a known good state before interacting with it. We can move this to a `Reset()` method, though. What do you think?

maruel replied:
Yes please. It's the last request after that I will merge the PR.

fd0 replied:
done.

fd0 commented:
Let me know if there's anything else.

maruel commented on experimental/devices/bmp180/bmp180.go:
```diff
@@ -0,0 +1,247 @@
+
+// Dev is a handle to a bmp180.
+type Dev struct {
+	dev *mmr.Dev8
```
I just realized you don't need it to be a pointer, please inline instead of allocating on the heap. Last comment I promise.

fd0 replied:
He, don't worry, when working with external contributors for [restic](https://github.com/restic/restic) (my project) I'm regularly on the other side of this.
